templates/verifier_groth16.vy.ejs

```diff
@@ -0,0 +1,189 @@
+    for i in range(4): 
+        input.append(convert(p1[i].X, bytes32))
+        input.append(convert(p1[i].Y, bytes32))
+        input.append(convert(p2[i].X[0], bytes32))
+        input.append(convert(p2[i].X[1], bytes32))
+        input.append(convert(p2[i].Y[0], bytes32))
+        input.append(convert(p2[i].Y[1], bytes32))
+
+    converted: Bytes[768] = concat(input[0], input[1], input[2], input[3], input[4],input[5], 
+                                   input[6], input[7], input[8], input[9], input[10], input[11],
+                                   input[12], input[13], input[14], input[15], input[16], input[17], 
+                                   input[18],input[19], input[20], input[21], input[22], input[23])
+
+    response: Bytes[32] = b''
+
+    # raw_call to ecpairing contract
+    response = raw_call(0x0000000000000000000000000000000000000008, 
+                        converted, 
+                        max_outsize=32, 
+                        is_static_call=True
+                        )
+
+    return _abi_decode(response, bool)
+
+
+# @dev compute verifying key
+@internal
+@pure
+def verifying_key() -> VerifyingKey:
+    vk: VerifyingKey = VerifyingKey(
+        {
+            alfa1: G1Point({X: <%=vk_alpha_1[0]%>, Y: <%=vk_alpha_1[1]%>}),
+            beta2: G2Point({X: [<%=vk_beta_2[0][1]%>, <%=vk_beta_2[0][0]%>], Y: [<%=vk_beta_2[1][1]%>, <%=vk_beta_2[1][0]%>]}), 
+            gamma2: G2Point({X: [<%=vk_gamma_2[0][1]%>, <%=vk_gamma_2[0][0]%>], Y: [<%=vk_gamma_2[1][1]%>, <%=vk_gamma_2[1][0]%>]}),
+            delta2: G2Point({X: [<%=vk_delta_2[0][1]%>, <%=vk_delta_2[0][0]%>], Y: [<%=vk_delta_2[1][1]%>, <%=vk_delta_2[1][0]%>]}),  
+            ic: []
+        }
+    )
+    <% for (let i=0; i<IC.length; i++) { %>
+    vk.ic.append(G1Point({X: <%=IC[i][0]%>, Y: <%=IC[i][1]%>}))
+    <% } %>
+    return vk
+
+
+# @dev Verifies the proof
+@external
+@view
+def verify_proof(a: uint256[2], b: uint256[2][2], c: uint256[2], input: DynArray[uint256, <%=IC.length-1%>]) -> bool:
+    # If the values are not in the correct range, the Pairing contract will revert.
+    proof: Proof = Proof({
+        A: G1Point({X: a[0], Y: a[1]}),
+        B: G2Point({X: [b[0][0], b[0][1]], Y: [b[1][0], b[1][1]]}),
+        C: G1Point({X: c[0], Y: c[1]})
+    })
+
+    vk: VerifyingKey = self.verifying_key()
+
+    # Compute the linear combination vk_x of inputs times IC
+    assert unsafe_add(len(input), 1) == len(vk.ic), "invalid proof"
+    vk_x: G1Point = vk.ic[0]
+
+    if N_INPUTS > 1:
+        for i in range(N_INPUTS):
+            vk_x = self.addition(vk_x, self.scalar_mul(vk.ic[i+1], input[i]))
+    else:
+        vk_x = self.addition(vk_x, self.scalar_mul(vk.ic[1], input[0]))
+
+    # check pairing
+    p1: DynArray[G1Point, 4] = [
+        self.negate(proof.A),
+        vk.alfa1,
+        vk_x,
+        proof.C
+    ]
+    p2: DynArray[G2Point, 4] = [
+        proof.B,
+        vk.beta2,
+        vk.gamma2,
+        vk.delta2
+    ]
```

Comment:
related to my comment about pairing_check_4 -- these would be better i think as static arrays.